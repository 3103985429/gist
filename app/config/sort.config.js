const SORT_ORDER_ASC = 'Ascending';
const SORT_ORDER_DESC = 'Descending';

module.exports = {
    SORT_ORDER_ASC,
    SORT_ORDER_DESC,
};
