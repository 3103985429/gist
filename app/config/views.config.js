const VIEW_TYPE_OBJECT_DETAILS = 'objectDetail';
const VIEW_TYPE_AREA_CHART = 'areaChart';
const VIEW_TYPE_BAR_CHART = 'barChart';
const VIEW_TYPE_BUBBLE_CHART_BASE = 'bubbleChart';
const VIEW_TYPE_BUBBLE_CHART = 'bubble-chart';
const VIEW_TYPE_GALLERY = 'gallery';
const VIEW_TYPE_GLOBE = 'globe';
const VIEW_TYPE_LINE_GRAPH = 'lineGraph';
const VIEW_TYPE_MAP = 'map';
const VIEW_TYPE_PIE_CHART = 'pieChart';
const VIEW_TYPE_PIE_SET = 'pieSet';
const VIEW_TYPE_REGIONAL_MAP = 'regional-map';
const VIEW_TYPE_SCATTERPLOT = 'scatterplot';
const VIEW_TYPE_TABLE = 'table';
const VIEW_TYPE_TIMELINE = 'timeline';
const VIEW_TYPE_GROUPED_GALLERY = 'grouped-gallery';
const VIEW_TYPE_TREEMAP = 'treemap';
const VIEW_TYPE_WORD_CLOUD = 'wordCloud';

const SETTING_DEFAULT_GROUP_BY_DURATION = 'defaultGroupByDuration';
const SETTING_DEFAULT_GROUP_BY_RANGE = 'defaultGroupByRange';
const SETTING_DEFAULT_GROUP_BY_COLUMN_NAME = 'defaultGroupByColumnName';
const SETTING_DEPENDS_ON_SORT_BY = 'dependsOnSortBy';

const GALLERY_STYLE_IMAGE = 'image';
const GALLERY_STYLE_ICON = 'icon';
const GALLERY_STYLE_HEATMAP = 'heatmap';

const MAP_STYLE_COUNTRY = 'country';
const MAP_STYLE_PIN = 'pin';
const MAP_STYLE_BUBBLE = 'bubble';
const MAP_STYLE_HEATMAP = 'heatmap';

module.exports = {
    VIEW_TYPE_OBJECT_DETAILS,
    VIEW_TYPE_AREA_CHART,
    VIEW_TYPE_BAR_CHART,
    VIEW_TYPE_BUBBLE_CHART_BASE,
    VIEW_TYPE_BUBBLE_CHART,
    VIEW_TYPE_GALLERY,
    VIEW_TYPE_GLOBE,
    VIEW_TYPE_LINE_GRAPH,
    VIEW_TYPE_MAP,
    VIEW_TYPE_PIE_CHART,
    VIEW_TYPE_PIE_SET,
    VIEW_TYPE_REGIONAL_MAP,
    VIEW_TYPE_SCATTERPLOT,
    VIEW_TYPE_TABLE,
    VIEW_TYPE_TIMELINE,
    VIEW_TYPE_GROUPED_GALLERY,
    VIEW_TYPE_TREEMAP,
    VIEW_TYPE_WORD_CLOUD,
    SETTING_DEFAULT_GROUP_BY_DURATION,
    SETTING_DEFAULT_GROUP_BY_RANGE,
    SETTING_DEFAULT_GROUP_BY_COLUMN_NAME,
    SETTING_DEPENDS_ON_SORT_BY,
    GALLERY_STYLE_IMAGE,
    GALLERY_STYLE_ICON,
    GALLERY_STYLE_HEATMAP,
    MAP_STYLE_COUNTRY,
    MAP_STYLE_PIN,
    MAP_STYLE_BUBBLE,
    MAP_STYLE_HEATMAP,
};
