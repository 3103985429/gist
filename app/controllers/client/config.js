// legacy
const { cond, constant, eq, kebabCase, stubTrue } = require('lodash');

const {
    VIEW_TYPE_BAR_CHART,
    VIEW_TYPE_PIE_SET,
    VIEW_TYPE_SCATTERPLOT,
    VIEW_TYPE_TIMELINE,
    VIEW_TYPE_TREEMAP,
} = require('../../config/views.config');

const PAGE_SIZE_BAR_CHART = 1000;
const PAGE_SIZE_PIE_SET = 50;
const PAGE_SIZE_SCATTERPLOT = 1000;
const PAGE_SIZE_TIMELINE = 20;
const PAGE_SIZE_TREEMAP = 1000;
const PAGE_SIZE_DEFAULT = 200;

function defaultPageSize(type) {
    const equals = a => b => eq(a, b);

    return cond([
        [equals(VIEW_TYPE_BAR_CHART), constant(PAGE_SIZE_BAR_CHART)],
        [equals(VIEW_TYPE_PIE_SET), constant(PAGE_SIZE_PIE_SET)],
        [equals(VIEW_TYPE_SCATTERPLOT), constant(PAGE_SIZE_SCATTERPLOT)],
        [equals(VIEW_TYPE_TIMELINE), constant(PAGE_SIZE_TIMELINE)],
        [equals(VIEW_TYPE_TREEMAP), constant(PAGE_SIZE_TREEMAP)],
        [stubTrue, constant(PAGE_SIZE_DEFAULT)],
    ])(type);
}

const numberOperations = ['ToInteger', 'ToFloat', 'ToPercent', 'ToCurrency'];

const groupSizes = {
    timeline: [19, 11],
    pieSet: 50,
};
const defaultDateFormat = 'MM/DD/YYYY';

const AGGREGATE_BY_DEFAULT_COLUMN_NAME = 'Number of Items';

const AGGREGATE_BY_DISABLED_COLUMN_NAME = 'None';

const GROUP_BY_DATE_DISABLED = 'None';

const isDate = (dataSourceDescription, columnName) => {
    if (dataSourceDescription.raw_rowObjects_coercionScheme) {
        const coercion = dataSourceDescription.raw_rowObjects_coercionScheme;
        if (coercion[columnName] && coercion[columnName].operation === 'ToDate') {
            return true;
        } else {
            // if there are any custom fields to process, recurse through them
            for (let i = 0; i < dataSourceDescription.customFieldsToProcess.length; i++) {
                const mergedFields = dataSourceDescription.customFieldsToProcess[i].fieldsToMergeIntoArray;
                const fieldName = dataSourceDescription.customFieldsToProcess[i].fieldName;
                if (fieldName === columnName) {
                    return isDate(dataSourceDescription, mergedFields[i]);
                }
            }
        }
    }
    return false;
};

const isNumber = (dataSourceDescription, columnName) => {
    if (dataSourceDescription.raw_rowObjects_coercionScheme) {
        const coercion = dataSourceDescription.raw_rowObjects_coercionScheme;
        if (coercion[columnName] && numberOperations.includes(coercion[columnName].operation)) {
            return true;
        }
    }
    return false;
};

const formatDefaultView = view => kebabCase(view);

const hasMoreThanOneView = views => {
    let numberOfViews = 0;
    for (let view in views) {
        if (views[view].visible === true) {
            numberOfViews++;
        }
        if (numberOfViews > 1) {
            return true;
        }
    }
    return false;
};

// todo: merge with calculatePageRanges?
const calculateGalleryPageRanges = numResults => {
    let pageRanges = [];

    if (numResults > 200) {
        pageRanges = [5, 10, 20, 50, 100, 200];
    } else if (numResults > 100) {
        pageRanges = [5, 10, 20, 50, 100];
    } else if (numResults > 50) {
        pageRanges = [5, 10, 20, 50];
    } else if (numResults > 20) {
        pageRanges = [5, 10, 20];
    } else if (numResults > 10) {
        pageRanges = [5, 10];
    } else if (numResults > 5) {
        pageRanges = [5];
    }

    if (!pageRanges.includes(numResults) && numResults < 200) {
        pageRanges.push(numResults);
    }

    return pageRanges;
};

const userRoles = {
    visualizationEditor: 'visualizationEditor',
    admin: 'admin',
    editor: 'editor',
    viewer: 'viewer',
};

const PERCENT_UNIT = '%';

module.exports = {
    defaultPageSize,
    groupSizes,
    defaultDateFormat,
    AGGREGATE_BY_DEFAULT_COLUMN_NAME,
    AGGREGATE_BY_DISABLED_COLUMN_NAME,
    GROUP_BY_DATE_DISABLED,
    PERCENT_UNIT,
    isDate,
    isNumber,
    formatDefaultView,
    hasMoreThanOneView,
    calculateGalleryPageRanges,
    userRoles,
};
