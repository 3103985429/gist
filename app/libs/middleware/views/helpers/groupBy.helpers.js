const { cond, constant, eq, has, stubTrue } = require('lodash');

const { RealColumnNameFromHumanReadableColumnName } = require('../../../datasources/imported_data_preparation');
const {
    SETTING_DEFAULT_GROUP_BY_DURATION,
    SETTING_DEFAULT_GROUP_BY_RANGE,
    SETTING_DEFAULT_GROUP_BY_COLUMN_NAME,
    SETTING_DEPENDS_ON_SORT_BY,
} = require('../../../../config/views.config');
const { fieldOverrideIfExists } = require('../../../../controllers/client/func');
const { equals } = require('../../../../utils/helpers');

const TYPE_DATE = 'date';
const TYPE_NUMBER = 'number';
const TYPE_STRING = 'string';

module.exports = {
    getDefaultGroupByColumnName,
    getGroupByTypeFromSortBy,
    invalidateGroupBy,
};

/**
 * @param {String} groupByType
 * @param {Object} dataSourceDescription
 * @param {String?} defaultGroupByDuration
 * @param {String?} defaultGroupByRange
 * @param {String?} defaultGroupByColumnName
 * @returns {String|undefined}
 */
function getDefaultGroupByColumnName(
    groupByType,
    dataSourceDescription,
    { defaultGroupByDuration, defaultGroupByRange, defaultGroupByColumnName }
) {
    if (groupByType === TYPE_DATE) {
        return defaultGroupByDuration;
    }

    if (groupByType === TYPE_NUMBER) {
        return defaultGroupByRange;
    }

    return fieldOverrideIfExists(defaultGroupByColumnName, dataSourceDescription);
}

/**
 * @param {Boolean} isDate
 * @param {Boolean} isNumber
 * @param {Object[]} masterViewSettings
 * @returns {String|null}
 */
function getGroupByTypeFromSortBy({ isDate, isNumber } = {}, masterViewSettings) {
    const findSetting = settingName => masterViewSettings.find(({ name }) => name === settingName);
    const getType = cond([
        [() => isDate && findSetting(SETTING_DEFAULT_GROUP_BY_DURATION), constant(TYPE_DATE)],
        [() => isNumber && findSetting(SETTING_DEFAULT_GROUP_BY_RANGE), constant(TYPE_NUMBER)],
        [
            () => has(findSetting(SETTING_DEFAULT_GROUP_BY_COLUMN_NAME), SETTING_DEPENDS_ON_SORT_BY),
            constant(TYPE_STRING),
        ],
        [stubTrue, constant(null)],
    ]);

    return getType();
}

function invalidateGroupBy(
    groupBy,
    groupByType,
    dataSourceDescription,
    { fieldsNotAvailableAsGroupByColumns = [], durationsAvailableForGroupBy = [], rangesAvailableForGroupBy = [] }
) {
    const { fe_excludeFields } = dataSourceDescription;
    const groupByRealColumnName = RealColumnNameFromHumanReadableColumnName(groupBy, dataSourceDescription, true);
    const stringGroupBy = () =>
        eq(fe_excludeFields[groupByRealColumnName], false) && !fieldsNotAvailableAsGroupByColumns[groupByRealColumnName]
            ? groupBy
            : undefined;
    const dateGroupBy = () => (durationsAvailableForGroupBy.includes(groupBy) ? groupBy : undefined);
    const numberGroupBy = () => (rangesAvailableForGroupBy.includes(groupBy) ? groupBy : undefined);
    const getDefaultGroupBy = cond([
        [equals(TYPE_STRING), stringGroupBy],
        [equals(TYPE_DATE), dateGroupBy],
        [equals(TYPE_NUMBER), numberGroupBy],
        [stubTrue, constant(groupBy)],
    ]);

    return getDefaultGroupBy(groupByType);
}
