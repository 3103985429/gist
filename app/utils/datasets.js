const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
