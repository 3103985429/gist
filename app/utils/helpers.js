const { isEqual, trim, toString } = require('lodash');

module.exports.equals = a => b => isEqual(a, b);

module.exports.isEmbed = (query) => query.embed === 'true' || query.embed === true;

module.exports.formatColumnName = (value) => trim(toString(value).replace(/\./g, '_'));

module.exports.formatRawValues = (value) => trim(toString(value).replace(/’/g, '\'').replace(/”/g, '"'));

module.exports.addSlugIndex = (slugs, item, tilteProp = 'title') => {
    if (!slugs.length) {
        return;
    }

    let i = 2;
    const slugList = slugs.map((item) => item.slug);
    const max = i + slugList.length + 1;

    for (i; i < max; i++) {
        const newSlug = `${item.slug}-${i}`;

        if (!slugList.includes(newSlug)) {
            item.slug = newSlug;
            item[tilteProp] = `${item[tilteProp]} ${i}`;

            break;
        }
    }
};
