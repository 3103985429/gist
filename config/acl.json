{
    "standardViz": [
        "seeFields",
        "seeFilters",
        "seeContent",
        "seeViews",
        "seeStandardViews",
        "seeSettings",
        "editFields",
        "editViews",
        "editGeneralSettings",
        "editObjectDetail",
        "useStandardImport",
        "useStandardSaveColumns",
        "useStandardGet"
    ],
    "superAdmin": [
        "createViz",
        "seeViz",
        "editViz",
        "manageTeams",
        "seeArticles",
        "createArticles",
        "editArticles",
        "manageAllArticles",
        "seeSites",
        "editSites",
        "createSites",
        "seeInsights",
        "editInsights",
        "createInsights",
        "seePerformance",
        "manageUsers"
    ],
    "admin": [
        "manageTeams",
        "createViz",
        "seeViz",
        "editViz",
        "seeArticles",
        "createArticles",
        "editArticles",
        "manageAllArticles",
        "seeInsights",
        "editInsights",
        "createInsights",
        "seePerformance",
        "manageUsers"
    ],
    "vizEditor": [
        "seeViz",
        "editViz",
        "seeInsights",
        "editInsights",
        "createInsights",
        "seePerformance"
    ],
    "vizViewer": [
        "seeViz",
        "seeInsights",
        "editInsights",
        "createInsights"
    ],
    "vizAuthor": [
        "createViz",
        "seeViz",
        "editViz",
        "seePerformance"
    ],
    "articleEditor": [
        "seeArticles",
        "editArticles"
    ],
    "articleViewer": [
        "seeArticles"
    ],
    "articleAuthor": [
        "createArticles",
        "seeArticles",
        "editArticles"
    ],
    "siteViewer": [
        "seeSites"
    ],
    "siteEditor": [
        "seeSites",
        "editSites"
    ],
    "siteAuthor": [
        "createSites",
        "seeSites",
        "editSites"
    ],
    "team": [
        "createStandardViz"
    ]
}
