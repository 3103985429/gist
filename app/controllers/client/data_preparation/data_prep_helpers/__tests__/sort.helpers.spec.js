import { pokemon1 } from '../../../../../../tests/config/datasets';
import { SORT_ORDER_ASC } from '../../../../../config/sort.config';
import {
    VIEW_TYPE_GALLERY,
    VIEW_TYPE_TIMELINE,
    VIEW_TYPE_BUBBLE_CHART,
    VIEW_TYPE_SCATTERPLOT,
    VIEW_TYPE_TABLE,
} from '../../../../../config/views.config';
import { compareCreatedAt, getGallerySort, getSort, getTimelineSort } from '../sort.helpers';


describe('sort helpers', () => {
    let dataSource;

    beforeEach(() => {
        dataSource = {};
        Object.assign(dataSource, pokemon1, {
            fe_views: {
                views: {
                    gallery: {
                        defaultSortByColumnName: 'Image URL',
                        defaultSortOrderDescending: false,
                        visible: true,
                    },
                    timeline: {
                        defaultSortByColumnName: 'Image URL',
                        defaultSortOrderDescending: true,
                        visible: true,
                    },
                    table: {
                        defaultSortByColumnName: 'Image URL',
                        defaultSortOrderDescending: true,
                        visible: true,
                    },
                },
                default_view: 'gallery',
            },
        });
    });

    describe('compareCreatedAt', () => {
        it('should return 0 when values are equal', () => {
            const result = compareCreatedAt({ createdAt: 1 }, { createdAt: 1 });
            expect(result).toBe(0);
        });

        it('should return 1 when first value is bigger', () => {
            const result = compareCreatedAt({ createdAt: 2 }, { createdAt: 1 });
            expect(result).toBe(1);
        });

        it('should return -1 when first value is smaller', () => {
            const result = compareCreatedAt({ createdAt: 0 }, { createdAt: 1 });
            expect(result).toBe(-1);
        });
    });

    describe('getSort', () => {
        it('should return sort for gallery', () => {
            const result = getSort(VIEW_TYPE_GALLERY, dataSource, { sortBy: 'Name' });
            expect(result).toMatchSnapshot();
        });

        it('should return sort for timeline', () => {
            const result = getSort(VIEW_TYPE_TIMELINE, dataSource, { sortBy: 'Name' });
            expect(result).toMatchSnapshot();
        });

        it('should return sort for bubble chart', () => {
            const result = getSort(VIEW_TYPE_BUBBLE_CHART, dataSource, {
                groupBy_realColumnName: 'Date',
                defaultXAxisField: 'Data 1',
                defaultYAxisField: 'Data 2',
            });
            expect(result).toMatchSnapshot();
        });

        it('should return sort for scatterplot', () => {
            const result = getSort(VIEW_TYPE_SCATTERPLOT, dataSource, {
                defaultXAxisField: 'Data 1',
                defaultYAxisField: 'Data 2',
            });
            expect(result).toMatchSnapshot();
        });

        it('should return sort for table', () => {
            const result = getSort(VIEW_TYPE_TABLE, dataSource, { sortBy: 'Name', sortDirection: 'Descending' });
            expect(result).toMatchSnapshot();
        });
    });

    describe('getGallerySort', () => {
        it('should return sort for gallery', () => {
            const result = getGallerySort(dataSource, { sortBy: 'Name', sortDirection: SORT_ORDER_ASC });
            expect(result).toMatchSnapshot();
        });
    });

    describe('getTimelineSort', () => {
        it('should return sort for timeline', () => {
            const result = getTimelineSort(dataSource, { sortBy: 'Name', sortDirection: SORT_ORDER_ASC });
            expect(result).toMatchSnapshot();
        });
    });
});
