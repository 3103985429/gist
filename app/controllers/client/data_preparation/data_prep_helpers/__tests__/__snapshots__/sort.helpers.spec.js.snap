// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort helpers getGallerySort should return sort for gallery 1`] = `
Object {
  "_id": 1,
  "rowParams.Name": 1,
  "size": -1,
}
`;

exports[`sort helpers getSort should return sort for bubble chart 1`] = `
Object {
  "rowParams.Data 1": 1,
  "rowParams.Data 2": 1,
  "rowParams.Date": 1,
}
`;

exports[`sort helpers getSort should return sort for gallery 1`] = `
Object {
  "_id": 1,
  "rowParams.Name": 1,
  "size": -1,
}
`;

exports[`sort helpers getSort should return sort for scatterplot 1`] = `
Object {
  "_id": 1,
  "rowParams.Data 1": 1,
  "rowParams.Data 2": 1,
}
`;

exports[`sort helpers getSort should return sort for table 1`] = `
Object {
  "rowParams.Name": -1,
}
`;

exports[`sort helpers getSort should return sort for timeline 1`] = `
Object {
  "_id": -1,
  "rowParams.Name": -1,
}
`;

exports[`sort helpers getTimelineSort should return sort for timeline 1`] = `
Object {
  "_id": 1,
  "rowParams.Name": 1,
}
`;
